File: client/src/boundary/interactor/BoardD2BConverter.test.ts
import { Board, Ship } from "domain/index";
import { BoundaryBoard, BoundaryShip } from "boundary/model/";
import { BoardD2BConverter } from "./BoardD2BConverter";
import { ShipD2BConverter } from "./ShipD2BConverter";
import { mock, MockProxy } from "jest-mock-extended";

describe(BoardD2BConverter, () => {
  let boardD2BConverter: BoardD2BConverter;
  let shipD2BConverter: MockProxy<ShipD2BConverter>;

  beforeEach(() => {
    shipD2BConverter = mock<ShipD2BConverter>();
    boardD2BConverter = new BoardD2BConverter(shipD2BConverter);
  });

  it("Converts Board model to BoundaryBoard model", () => {
    const ships: Ship[] = [mock<Ship>()];
    const expectedShips: BoundaryShip[] = [mock<BoundaryShip>()];
    const inputBoard: Board = new Board([[0, 0, 0, 0]], ships);
    const expectedBoard: BoundaryBoard = new BoundaryBoard(
      [[0, 0, 0, 0]],
      expectedShips
    );

    shipD2BConverter.convertAll
      .calledWith(ships)
      .mockReturnValue(expectedShips);

    const convertedBoard: BoundaryBoard = boardD2BConverter.convert(inputBoard);

    expect(convertedBoard.grid).toEqual(expectedBoard.grid);
    expect(convertedBoard.ships).toEqual(expectedBoard.ships);
  });
});

File: server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/BoardD2BConverterTest.java
package eu.optas.use_cases.implementation;

import eu.optas.domain.Board;
import eu.optas.domain.Ship;
import eu.optas.use_cases.api.BoundaryBoard;
import eu.optas.use_cases.api.BoundaryShip;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class BoardD2BConverterTest {

    @Test
    void convert() {
        ShipD2BConverter shipD2BConverterMock = mock(ShipD2BConverter.class);
        BoardD2BConverter boardD2BConverter = new BoardD2BConverter(shipD2BConverterMock);

        List<List<Integer>> grid = List.of(List.of(0, 0, 0, 0));

        List<Ship> ships = new ArrayList<>();
        List<BoundaryShip> boundaryShips = new ArrayList<>();
        Board inputBoard = new Board(grid, ships);
        BoundaryBoard expectedBoard = new BoundaryBoard(grid, boundaryShips);

        when(shipD2BConverterMock.convertAll(ships)).thenReturn(boundaryShips);

        BoundaryBoard convertedBoard = boardD2BConverter.convert(inputBoard);

        assertThat(convertedBoard).usingRecursiveComparison().isEqualTo(expectedBoard);
    }
}